OSSM-8952: Cluster-wide migration using the istio revision label (#89164)
openshift/openshift-docs: modules/ossm-migrating-workloads-using-the-istio-revision-label.adoc

// Module included in the following assemblies:
//
// * service-mesh-docs-main/migrating/checklists/ossm-migrating-cluster-wide-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-migrating-workloads-using-the-istio-revision-label_{context}"]
= Migrating workloads using the Istio revision label

Now you can migrate your workloads from the {SMProduct} 2.6 control plane to the {SMproduct} 3.0 control plane.

> Revision tags are not used in this example for simplicity. When migrating large meshes, you can use revision tags to avoid re-labeling all namespaces during future 3.y.z upgrades.

Reviewer: ```suggestion
Revision tags are not used in this example for simplicity. When migrating large meshes, you can use revision tags to avoid re-labeling all namespaces during future version 3 updates.
```
See https://www.ibm.com/docs/en/ibm-style?topic=references-product-versions#version-release-and-modification-identifiers
https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/term_glossary.adoc#update

Author: done

[NOTE]
====
You can migrate workloads and gateways separately, and in any order. For more information, see "Migrating gateways".
====

.Procedure

> . Find the current `IstioRevision` for your {SMProduct} 3.0 control plane by running the following command:

Author: done

+
[source,terminal]
----

> $ oc get istios istio-tenant-a

Reviewer 2: ```suggestion
$ oc get istios
```

Author: done

----
+
.Example output
+
[source,terminal]
----
NAME             REVISIONS   READY   IN USE   ACTIVE REVISION   STATUS    VERSION   AGE

> istio-tenant-a   1           1       0        istio-tenant-a    Healthy   v1.24.3   30s

Reviewer 2: ```suggestion
ossm-3-v1-24-3   1           1       0        ossm-3-v1-24-3    Healthy   v1.24.3   30s
```

Reviewer 2: I will check

Reviewer 2: You are right. I was too fast with manual edits without testing.

Author: done

----

> . Copy the "ACTIVE REVISION" to use as your `istio.io/rev` label in the next step.

Reviewer: ```suggestion
. Copy the value in the `ACTIVE REVISION` column to use as your `istio.io/rev` label in the next step.
```

Author: Done

====
The naming format of your revisions depends on which upgrade strategy you choose for your `Istio` instance.
====

. Update the injection labels on the `dataplane` namespace by running the following command:
+
[source,terminal]
----
$ oc label ns bookinfo istio.io/rev=ossm-3-v1-24-3 maistra.io/ignore-namespace="true" istio-injection- --overwrite=true
----
+

> The command adds the following labels to the namespace:

Reviewer 2: This should be included in previous step. Currently it's listed as new step but it refers the command in previous step. So there should be just one step which: adds two labels and removes one

+
Running the command performs the following actions:

> .. Removes the `istio-injection` label: This label prevents the 3.0 control plane from injecting the proxy. The `istio-injection` label takes precedence over `istio.io/rev` label. Because you cannot create the default `IstioRevisionTag` yet, you must temporarily remove the `istio-injection=enabled`. Leaving the `istio-injection=enabled` label applied would prevent the 3.0 control plane from performing proxy injection. 

Author: Thank you, these are some of the items I wanted to discuss.

.. Adds the `istio.io/rev=ossm-3-v1-24-3` label: This label ensures that any newly created or restarted pods in the namespace connect to the {SMProduct} 3.0 proxy. 

> .. Adds the `maistra.io/ignore-namespace: "true"` label: This label disables sidecar injection for {SMProduct} 2.6 proxies in the namespace. With the label applied, {SMProduct} 2.6 stops injecting proxies in this namespace, and any new proxies are injected by {SMProduct} 3.0. Without this label, the {SMProduct} 2.6 injection webhook tries to inject the pod and the injected sidecar proxy will refuse to start since it will have both the {SMProduct} 2.6 and the {SMProduct} 3.0 Container Network Interface(CNI) annotations.

Reviewer: ```suggestion
.. Adds the `maistra.io/ignore-namespace: "true"` label: This label disables sidecar injection for {SMProduct} 2.6 proxies in the namespace. With the label applied, {SMProduct} 2.6 stops injecting proxies in this namespace, and any new proxies are injected by {SMProduct} 3.0. Without this label, the {SMProduct} 2.6 injection webhook tries to inject the pod and the injected sidecar proxy refuses to start since it will have both the {SMProduct} 2.6 and the {SMProduct} 3.0 Container Network Interface(CNI) annotations.
```
Avoid future tense: https://www.ibm.com/docs/en/ibm-style?topic=grammar-verbs#tense

Author: Done

+
.Example command for `bookinfo` application
[source,terminal]
----
$ oc rollout restart deployments -n bookinfo
----

.. To restart each workload individually, run the following command for each workload:
+
.Example command with `bookinfo` application
[source,terminal]
----
$ oc rollout restart deployments productpage-v1 -n bookinfo
----

. Wait for the `productpage` application to restart by running the following command:
+
[source,terminal]
----
$ oc rollout status deployment productpage-v1 -n bookinfo
----

.Verification

. Ensure that expected workloads are managed by the new control plane by running the following command:
+
[source,terminal]
----
$ istioctl ps -n bookinfo
----
+
.Sample output:
[source,terminal]
----
$ istioctl ps -n bookinfo
NAME                                          CLUSTER        CDS             LDS             EDS             RDS             ECDS         ISTIOD                                           VERSION
details-v1-7f46897b-d497c.bookinfo            Kubernetes     SYNCED          SYNCED          SYNCED          SYNCED          NOT SENT     istiod-install-istio-system-866b57d668-6lpcr     1.20.8

> productpage-v1-74bfbd4d65-vsxqm.bookinfo      Kubernetes     SYNCED (4s)     SYNCED (4s)     SYNCED (3s)     SYNCED (4s)     IGNORED      istiod-ossm-3-v1-24-1-797bb4d78f-xpchx           1.24.3

Author: done

// Module included in the following assemblies:
//
// * service-mesh-docs-main/migrating/checklists/ossm-migrating-cluster-wide-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-migrating-workloads-using-the-istio-revision-label_{context}"]
= Migrating workloads using the Istio revision label

Now you can migrate your workloads from the {SMProduct} 2.6 control plane to the {SMproduct} 3.0 control plane.

> [NOTE]

Reviewer 2: I would probably add another note about revision tags, something like:
Revision tags are not used in this example for simplicity but it's recommended to use them for big meshes to avoid re-labeling of namespaces during future 3.y.z upgrades. See <revision tags doc link>

Author: done

+
.Example command for `bookinfo` application
[source,terminal]
----
$ oc rollout restart deployments -n bookinfo
----

.. To restart each workload individually, run the following command for each workload:
+
.Example command with `bookinfo` application
[source,terminal]
----
$ oc rollout restart deployments productpage-v1 -n bookinfo
----

. Wait for the `productpage` application to restart by running the following command:
+
[source,terminal]
----
$ oc rollout status deployment productpage-v1 -n bookinfo
----

.Verification

. Ensure that expected workloads are managed by the new control plane by running the following command:
+
[source,terminal]
----
$ istioctl ps -n bookinfo
----
+
.Sample output:
[source,terminal]
----
$ istioctl ps -n bookinfo
NAME                                          CLUSTER        CDS             LDS             EDS             RDS             ECDS         ISTIOD                                           VERSION
details-v1-7f46897b-d497c.bookinfo            Kubernetes     SYNCED          SYNCED          SYNCED          SYNCED          NOT SENT     istiod-install-istio-system-866b57d668-6lpcr     1.20.8

> productpage-v1-74bfbd4d65-vsxqm.bookinfo      Kubernetes     SYNCED (4s)     SYNCED (4s)     SYNCED (3s)     SYNCED (4s)     IGNORED      istiod-ossm-3-v1-24-1-797bb4d78f-xpchx           1.24.1

Author: done

ratings-v1-559b64556-c5ppg.bookinfo           Kubernetes     SYNCED          SYNCED          SYNCED          SYNCED          NOT SENT     istiod-install-istio-system-866b57d668-6lpcr     1.20.8
reviews-v1-847fb7c54d-qxt5d.bookinfo          Kubernetes     SYNCED          SYNCED          SYNCED          SYNCED          NOT SENT     istiod-install-istio-system-866b57d668-6lpcr     1.20.8
reviews-v2-5c7ff5b77b-8jbhd.bookinfo          Kubernetes     SYNCED          SYNCED          SYNCED          SYNCED          NOT SENT     istiod-install-istio-system-866b57d668-6lpcr     1.20.8
reviews-v3-5c5d764c9b-rrx8w.bookinfo          Kubernetes     SYNCED          SYNCED          SYNCED          SYNCED          NOT SENT     istiod-install-istio-system-866b57d668-6lpcr     1.20.8
----
+

> Even if there are different versions of the proxies communication between services should work.

Reviewer 2: ```suggestion
The output above shows that only `productpage-v1` deployment has been restarted and was injected with 3.0 proxy. Even if there are different versions of the proxies communication between services still works.
```

> . If the 2.6 installation contains additional namespaces, migrate the next namespace now.

Reviewer 2: ```suggestion
. If the 2.6 installation contains additional data plane namespaces, migrate the next namespace now.
```

Author: done

// Module included in the following assemblies:
//
// * service-mesh-docs-main/migrating/checklists/ossm-migrating-cluster-wide-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="ossm-migrating-workloads-using-the-istio-revision-label_{context}"]
= Migrating workloads using the Istio revision label

Now you can migrate your workloads from the {SMProduct} 2.6 control plane to the {SMproduct} 3.0 control plane.

Revision tags are not used in this example for simplicity. When migrating large meshes, you can use revision tags to avoid re-labeling all namespaces during future 3.y.z upgrades.

[NOTE]
====
You can migrate workloads and gateways separately, and in any order. For more information, see "Migrating gateways".
====

.Procedure

. Update the injection labels on the data plane namespace by running the following command:
+
[source,terminal]
----

> $ oc label ns bookinfo istio.io/rev=ossm-3-v1-24-3 maistra.io/ignore-namespace="true" istio-injection- --overwrite=true

Reviewer 3: It would be good to include the step that tells users where to get the revision for the `istio.io/rev`. See steps 1 and 2 for the MultiTenant migration: https://89164--ocpdocs-pr.netlify.app/openshift-service-mesh/latest/migrating/multitenant/ossm-migrating-multitenant-assembly#migrating-multitenant-workloads_ossm-migrating-multitenant-assembly

Author: Done
